#!/bin/bash
#SBATCH --job-name=bam_qualimap
#SBATCH --cluster=genius
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=8       
#SBATCH --time=48:00:00         
#SBATCH --mem=32G               
#SBATCH --output=/scratch/leuven/362/vsc36201/RNA-seq/logs/qualimap_%j.out
#SBATCH --error=/scratch/leuven/362/vsc36201/RNA-seq/logs/qualimap_%j.err
#SBATCH -A lp_svbelleghem       

set -e  # Exit on error

module load cluster/genius/amd
module load Qualimap/2.3-foss-2022b-R-4.2.2
module load Java

# Define the directory containing BAM files
bam_dir="/scratch/leuven/362/vsc36201/RNA-seq/alignment"

# Create output directories if they don't exist
mkdir -p "$bam_dir"

# Loop through all BAM files generated by STAR
for bam_file in "${bam_dir}"/*.filtered.sorted.bam; do
    sample_name=$(basename "$bam_file" .filtered.sorted.bam)
    qualimap_dir="${bam_dir}/qualimap_${sample_name}"

    if [ -d "$qualimap_dir" ]; then
        echo "Qualimap already completed for $sample_name. Skipping." >> Progress_pipeline
    else
        echo "Running Qualimap on $sample_name..."
        qualimap bamqc -bam "$bam_file" --java-mem-size=8G -outdir "$qualimap_dir"
        echo "Qualimap QC completed for $sample_name." >> Progress_pipeline
    fi
done

echo "Qualimap analysis completed for all BAM files."
